File: app/src/main/java/it/cammino/risuscito/services/XmlImportService.java
package it.cammino.risuscito.services;

import android.annotation.TargetApi;
import android.app.IntentService;
import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.content.ContentResolver;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Build;
import android.support.v4.app.NotificationCompat;
import android.support.v4.app.NotificationManagerCompat;
import android.util.Log;
import android.util.Xml;

import com.crashlytics.android.Crashlytics;

import org.xmlpull.v1.XmlPullParser;
import org.xmlpull.v1.XmlPullParserException;

import java.io.IOException;
import java.io.InputStream;

import it.cammino.risuscito.LUtils;
import it.cammino.risuscito.ListaPersonalizzata;
import it.cammino.risuscito.R;
import it.cammino.risuscito.database.RisuscitoDatabase;
import it.cammino.risuscito.database.dao.ListePersDao;
import it.cammino.risuscito.database.entities.ListaPers;

public class XmlImportService extends IntentService {

  public static final String ACTION_URL = "it.cammino.risuscito.import.action.URL";
  public static final String ACTION_FINISH = "it.cammino.risuscito.import.action.URL";
  private static final String CHANNEL_ID = "itcr_import_channel";
  // We don't use namespaces
  private static final String ns = null;
  final int NOTIFICATION_ID = 2;
  final String TAG = getClass().getCanonicalName();

  public XmlImportService() {
    super("XmlImportService");
  }

  /**
   * This method is invoked on the worker thread with a request to process. Only one Intent is
   * processed at a time, but the processing happens on a worker thread that runs independently from
   * other application logic. So, if this code takes a long time, it will hold up other requests to
   * the same IntentService, but it will not hold up anything else. When all requests have been
   * handled, the IntentService stops itself, so you should not call {@link #stopSelf}.
   *
   * @param intent The value passed to {@link Context#startService(Intent)}.
   */
  @Override
  protected void onHandleIntent(Intent intent) {
    Log.d(TAG, "onHandleIntent: Starting");
    Uri data = intent.getData();
    if (data != null) {
      intent.setData(null);
      importData(data);
    }
  }

  private void importData(Uri data) {
    Log.d(TAG, "importData: data = " + data.toString());
    Log.d(TAG, "importData:  data.getScheme = " + data.getScheme());
    final String scheme = data.getScheme();

    NotificationManagerCompat mNotificationManager = NotificationManagerCompat.from(this);
    mNotificationManager.cancelAll();
    Notification mNotification;

    if (LUtils.hasO()) createChannel();

    mNotification =
        new NotificationCompat.Builder(this, CHANNEL_ID)
            .setSmallIcon(android.R.drawable.stat_sys_download)
            .setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
            .setContentTitle(getString(R.string.app_name))
            .setProgress(0, 0, true)
            .setContentText(getString(R.string.import_running))
            .build();

    mNotificationManager.notify(NOTIFICATION_ID, mNotification);

    if (ContentResolver.SCHEME_CONTENT.equals(scheme)) {
      try {
        InputStream fis = getContentResolver().openInputStream(data);
        ListaPersonalizzata celebrazione = parse(fis);

        if (celebrazione != null) {
          //                    DatabaseCanti listaCanti = new DatabaseCanti(this);
          //
          //                    SQLiteDatabase db = listaCanti.getReadableDatabase();
          //
          //                    ContentValues values = new ContentValues();
          //                    values.put("titolo_lista", celebrazione.getName());
          //                    values.put("lista",
          // ListaPersonalizzata.serializeObject(celebrazione));
          //
          //                    db.insert("LISTE_PERS", "", values);
          //
          //                    db.close();

          ListePersDao mDao = RisuscitoDatabase.getInstance(this).listePersDao();
          ListaPers listaPers = new ListaPers();
          listaPers.titolo = celebrazione.getName();
          listaPers.lista = celebrazione;
          mDao.insertLista(listaPers);

          mNotification =
              new NotificationCompat.Builder(this, CHANNEL_ID)
                  .setSmallIcon(R.drawable.ic_stat_action_done)
                  .setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
                  .setContentTitle(getString(R.string.app_name))
                  .setContentText(getString(R.string.import_done))
                  .build();

          mNotificationManager.notify(NOTIFICATION_ID, mNotification);

          Log.d(TAG, "Sending broadcast notification: ACTION_FINISH");
          Intent intentBroadcast = new Intent(ACTION_FINISH);
          sendBroadcast(intentBroadcast);

          Intent i =
              getBaseContext()
                  .getPackageManager()
                  .getLaunchIntentForPackage(getBaseContext().getPackageName());
          if (i != null) {
            i.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
          }
          startActivity(i);
          stopSelf();
        } else {
          mNotification =
              new NotificationCompat.Builder(this, CHANNEL_ID)
                  .setSmallIcon(R.drawable.ic_stat_alert_error)
                  .setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
                  .setContentTitle(getString(R.string.app_name))
                  .setTicker(getString(R.string.import_error))
                  .setContentText(getString(R.string.import_error))
                  .build();
          mNotificationManager.notify(NOTIFICATION_ID, mNotification);

          Log.d(TAG, "Sending broadcast notification: ACTION_FINISH");
          Intent intentBroadcast = new Intent(ACTION_FINISH);
          sendBroadcast(intentBroadcast);

          stopSelf();
        }
      } catch (XmlPullParserException | SecurityException | IOException e) {
        Log.e(TAG, "importData: " + e.getLocalizedMessage(), e);
        //        FirebaseCrash.log("importData: " + e.getMessage());
        Crashlytics.logException(e);
        mNotification =
            new NotificationCompat.Builder(this, CHANNEL_ID)
                .setSmallIcon(R.drawable.ic_stat_alert_error)
                .setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
                .setContentTitle(getString(R.string.app_name))
                .setTicker(getString(R.string.import_error))
                .setContentText(getString(R.string.import_error))
                .build();
        mNotificationManager.notify(NOTIFICATION_ID, mNotification);

        Log.d(TAG, "Sending broadcast notification: ACTION_FINISH");
        Intent intentBroadcast = new Intent(ACTION_FINISH);
        sendBroadcast(intentBroadcast);

        stopSelf();
      }
    }
  }

  public ListaPersonalizzata parse(InputStream in) throws XmlPullParserException, IOException {
    try {
      XmlPullParser parser = Xml.newPullParser();
      parser.setFeature(XmlPullParser.FEATURE_PROCESS_NAMESPACES, false);
      parser.setInput(in, null);
      parser.nextTag();
      return readLista(parser);
    } finally {
      in.close();
    }
  }

  private ListaPersonalizzata readLista(XmlPullParser parser)
      throws XmlPullParserException, IOException {
    ListaPersonalizzata list = new ListaPersonalizzata();
    Position tempPos;

    parser.require(XmlPullParser.START_TAG, ns, "list");
    String title = parser.getAttributeValue(null, "title");
    if (title != null) list.setName(parser.getAttributeValue(null, "title"));
    else {
      Log.e(TAG, "readLista: title is null");
      //      FirebaseCrash.log("importData: title is null");
      Crashlytics.log("importData: title is null");
      return null;
    }
    while (parser.next() != XmlPullParser.END_TAG) {
      if (parser.getEventType() != XmlPullParser.START_TAG) {
        continue;
      }
      String name = parser.getName();
      // Starts by looking for the entry tag
      if (name.equals("position")) {
        tempPos = readPosition(parser);
        list.addPosizione(tempPos.getName());
        if (!tempPos.getCanto().equalsIgnoreCase("0"))
          list.addCanto(tempPos.getCanto(), list.getNumPosizioni() - 1);
      } else {
        skip(parser);
      }
    }
    return list;
  }

  // Processes positions tags in the list.
  private Position readPosition(XmlPullParser parser) throws IOException, XmlPullParserException {
    Position result = new Position();
    parser.require(XmlPullParser.START_TAG, ns, "position");
    String name = parser.getAttributeValue(null, "name");
    String canto = readCanto(parser);
    parser.require(XmlPullParser.END_TAG, ns, "position");
    result.setName(name.trim());
    result.setCanto(canto.trim());
    return result;
  }

  // For the tags title and summary, extracts their text values.
  private String readCanto(XmlPullParser parser) throws IOException, XmlPullParserException {
    String result = "";
    if (parser.next() == XmlPullParser.TEXT) {
      result = parser.getText();
      parser.nextTag();
    }
    return result;
  }

  private void skip(XmlPullParser parser) throws XmlPullParserException, IOException {
    if (parser.getEventType() != XmlPullParser.START_TAG) {
      throw new IllegalStateException();
    }
    int depth = 1;
    while (depth != 0) {
      switch (parser.next()) {
        case XmlPullParser.END_TAG:
          depth--;
          break;
        case XmlPullParser.START_TAG:
          depth++;
          break;
      }
    }
  }

  @TargetApi(Build.VERSION_CODES.O)
  private void createChannel() {
    NotificationManager mNotificationManager =
        (NotificationManager)
            getApplicationContext().getSystemService(Context.NOTIFICATION_SERVICE);
    // The id of the channel.
    //        String id = CHANNEL_ID;
    // The user-visible name of the channel.
    CharSequence name = "XML Import";
    // The user-visible description of the channel.
    String description = "Importing selected XML";
    NotificationChannel mChannel =
        new NotificationChannel(CHANNEL_ID, name, NotificationManager.IMPORTANCE_DEFAULT);
    // Configure the notification channel.
    mChannel.setDescription(description);
    mChannel.setShowBadge(false);
    mChannel.setLockscreenVisibility(Notification.VISIBILITY_PUBLIC);
    if (mNotificationManager != null) {
      mNotificationManager.createNotificationChannel(mChannel);
    }
  }

  private static class Position {
    String name;
    String canto;

    public String getName() {
      return name;
    }

    public void setName(String name) {
      this.name = name;
    }

    public String getCanto() {
      return canto;
    }

    public void setCanto(String canto) {
      this.canto = canto;
    }
  }
}

File: app/src/main/java/it/cammino/risuscito/services/XmlImportService.kt
package it.cammino.risuscito.services

import android.app.Notification
import android.content.ContentResolver
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.util.Log
import android.util.Xml
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationManagerCompat
import androidx.work.Worker
import androidx.work.WorkerParameters
import com.google.firebase.crashlytics.ktx.crashlytics
import com.google.firebase.ktx.Firebase
import it.cammino.risuscito.ListaPersonalizzata
import it.cammino.risuscito.R
import it.cammino.risuscito.utils.Utility
import it.cammino.risuscito.database.RisuscitoDatabase
import it.cammino.risuscito.database.entities.ListaPers
import it.cammino.risuscito.utils.StringUtils
import org.xmlpull.v1.XmlPullParser
import org.xmlpull.v1.XmlPullParserException
import java.io.IOException
import java.io.InputStream

class XmlImportService(appContext: Context, workerParams: WorkerParameters) :
    Worker(appContext, workerParams) {

    override fun doWork(): Result {
        Log.d(TAG, "doWork: Starting")
        val resourceUri = inputData.getString(TAG_IMPORT_DATA)
        return importData(Uri.parse(resourceUri))
    }

    private fun importData(data: Uri): Result {
        Log.d(TAG, "$TAG_IMPORT_DATA: data = $data")
        Log.d(TAG, "$TAG_IMPORT_DATA:  data.getScheme = ${data.scheme}")
        val scheme = data.scheme
        val appContext = applicationContext


        val mNotificationManager = NotificationManagerCompat.from(appContext)
        mNotificationManager.cancelAll()
        var mNotification: Notification

        Utility.createNotificationChannelWrapper(
            appContext,
            CHANNEL_ID,
            "XML Import",
            "Importing selected XML"
        )

        mNotification = NotificationCompat.Builder(appContext, CHANNEL_ID)
            .setSmallIcon(android.R.drawable.stat_sys_download)
            .setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
            .setContentTitle(appContext.getString(R.string.app_name))
            .setProgress(0, 0, true)
            .setContentText(appContext.getString(R.string.import_running))
            .build()

        mNotificationManager.notify(NOTIFICATION_ID, mNotification)

        if (ContentResolver.SCHEME_CONTENT == scheme) {
            try {
                val fis = appContext.contentResolver.openInputStream(data)
                val celebrazione = parse(fis)

                val mDao = RisuscitoDatabase.getInstance(appContext).listePersDao()
                val listaPers = ListaPers()
                listaPers.titolo = celebrazione.name
                listaPers.lista = celebrazione
                mDao.insertLista(listaPers)

                mNotification = NotificationCompat.Builder(appContext, CHANNEL_ID)
                    .setSmallIcon(R.drawable.ic_notification_done)
                    .setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
                    .setContentTitle(appContext.getString(R.string.app_name))
                    .setContentText(appContext.getString(R.string.import_done))
                    .build()

                mNotificationManager.notify(NOTIFICATION_ID, mNotification)

                val i = appContext
                    .packageManager
                    .getLaunchIntentForPackage(appContext.packageName)
                i?.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
                appContext.startActivity(i)
                return Result.success()
            } catch (e: XmlPullParserException) {
                Log.e(TAG, TAG_IMPORT_DATA, e)
                Firebase.crashlytics.recordException(e)
                mNotification = NotificationCompat.Builder(appContext, CHANNEL_ID)
                    .setSmallIcon(R.drawable.ic_notification_error)
                    .setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
                    .setContentTitle(appContext.getString(R.string.app_name))
                    .setTicker(appContext.getString(R.string.import_error))
                    .setContentText(appContext.getString(R.string.import_error))
                    .build()
                mNotificationManager.notify(NOTIFICATION_ID, mNotification)

                return Result.failure()
            } catch (e: SecurityException) {
                Log.e(TAG, TAG_IMPORT_DATA, e)
                Firebase.crashlytics.recordException(e)
                mNotification = NotificationCompat.Builder(appContext, CHANNEL_ID)
                    .setSmallIcon(R.drawable.ic_notification_error)
                    .setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
                    .setContentTitle(appContext.getString(R.string.app_name))
                    .setTicker(appContext.getString(R.string.import_error))
                    .setContentText(appContext.getString(R.string.import_error)).build()
                mNotificationManager.notify(NOTIFICATION_ID, mNotification)

                return Result.failure()
            } catch (e: IOException) {
                Log.e(TAG, TAG_IMPORT_DATA, e)
                Firebase.crashlytics.recordException(e)
                mNotification = NotificationCompat.Builder(appContext, CHANNEL_ID)
                    .setSmallIcon(R.drawable.ic_notification_error)
                    .setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
                    .setContentTitle(appContext.getString(R.string.app_name))
                    .setTicker(appContext.getString(R.string.import_error))
                    .setContentText(appContext.getString(R.string.import_error)).build()
                mNotificationManager.notify(NOTIFICATION_ID, mNotification)

                return Result.failure()
            }

        }
        return Result.success()
    }

    @Throws(XmlPullParserException::class, IOException::class)
    private fun parse(inputStream: InputStream?): ListaPersonalizzata {
        inputStream.use {
            val parser = Xml.newPullParser()
            parser.setFeature(XmlPullParser.FEATURE_PROCESS_NAMESPACES, false)
            parser.setInput(it, null)
            parser.nextTag()
            return readLista(parser)
        }
    }

    @Throws(XmlPullParserException::class, IOException::class)
    private fun readLista(parser: XmlPullParser): ListaPersonalizzata {
        val list = ListaPersonalizzata()
        var tempPos: Position

        parser.require(XmlPullParser.START_TAG, ns, "list")
        val title = parser.getAttributeValue(null, "title")
        if (title != null)
            list.name = parser.getAttributeValue(null, "title")
        else {
            Log.e(TAG, "readLista: title is null")
            Firebase.crashlytics.log("$TAG_IMPORT_DATA: title is null")
        }
        while (parser.next() != XmlPullParser.END_TAG) {
            if (parser.eventType != XmlPullParser.START_TAG) {
                continue
            }
            val name = parser.name
            // Starts by looking for the entry tag
            if (name == POSITION_TAG) {
                tempPos = readPosition(parser)
                list.addPosizione(tempPos.name)
                if (!tempPos.canto.equals("0", ignoreCase = true))
                    list.addCanto(tempPos.canto, list.numPosizioni - 1)
            } else {
                skip(parser)
            }
        }
        return list
    }

    // Processes positions tags in the list.
    @Throws(IOException::class, XmlPullParserException::class)
    private fun readPosition(parser: XmlPullParser): Position {
        val result = Position()
        parser.require(XmlPullParser.START_TAG, ns, POSITION_TAG)
        val name = parser.getAttributeValue(null, "name")
        val canto = readCanto(parser)
        parser.require(XmlPullParser.END_TAG, ns, POSITION_TAG)
        result.name = name.trim { it <= ' ' }
        result.canto = canto.trim { it <= ' ' }
        return result
    }

    // For the tags title and summary, extracts their text values.
    @Throws(IOException::class, XmlPullParserException::class)
    private fun readCanto(parser: XmlPullParser): String {
        var result = StringUtils.EMPTY
        if (parser.next() == XmlPullParser.TEXT) {
            result = parser.text
            parser.nextTag()
        }
        return result
    }

    @Throws(XmlPullParserException::class, IOException::class)
    private fun skip(parser: XmlPullParser) {
        check(parser.eventType == XmlPullParser.START_TAG)
        var depth = 1
        while (depth != 0) {
            when (parser.next()) {
                XmlPullParser.END_TAG -> depth--
                XmlPullParser.START_TAG -> depth++
            }
        }
    }

    private class Position {
        lateinit var name: String
        lateinit var canto: String
    }

    companion object {
        internal const val NOTIFICATION_ID = 2
        internal val TAG = XmlImportService::class.java.canonicalName
        internal const val TAG_IMPORT_DATA = "importData"

        //        private const val JOB_ID = 5000
//        const val ACTION_URL = "it.cammino.risuscito.import.action.URL"
//        const val ACTION_FINISH = "it.cammino.risuscito.import.action.URL"
        private const val CHANNEL_ID = "itcr_import_channel"
        private const val POSITION_TAG = "position"

        // We don't use namespaces
        private val ns: String? = null

    }


}
